### PR TITLE
Guard against shape attributes that are not sequences

### LINKED ISSUE #1151: Array astype breaks when given a type instead of a dtype object
The code below fails in dask 0.9.0, numpy 1.11.0, toolz 0.7.4, Python 2.7:
```py
import numpy as np
import dask.array as da

a = np.arange(5).astype(np.int32)
d = da.from_array(a, (1,))
print(d.dtype)
e = d.astype(np.int16)
print(e.dtype)
print(e.compute())
```
with the exception
```
Traceback (most recent call last):
  File "./dask-astype.py", line 8, in <module>
    e = d.astype(np.int16)
  File "/home/bmerry/work/sdp/env/local/lib/python2.7/site-packages/dask/array/core.py", line 1091, in astype
    return elemwise(_astype, self, dtype, name=name)
  File "/home/bmerry/work/sdp/env/local/lib/python2.7/site-packages/dask/array/core.py", line 2159, in elemwise
    out_ndim = len(broadcast_shapes(*shapes))   # Raises ValueError if dimensions mismatch
  File "/home/bmerry/work/sdp/env/local/lib/python2.7/site-packages/dask/array/core.py", line 2131, in broadcast_shapes
    for sizes in zip_longest(*map(reversed, shapes), fillvalue=1):
TypeError: type object argument after * must be a sequence, not itertools.imap
```
If one passes `np.dtype(np.int16)` instead of just `np.int16`, then it works. The problem seems to be that the `np.int16.shape` attribute exists, but isn't a sequence. Possibly this can be fixed by adding `dtype = np.dtype(dtype) to the start of `Array.astype`.

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #1151 

cc @bmerry

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -2062,7 +2062,7 @@ def asarray(array):
     """
     if not isinstance(array, Array):
         name = 'asarray-' + tokenize(array)
-        if not hasattr(array, 'shape'):
+        if isinstance(getattr(array, 'shape', None), Iterable):
```
Comment (shoyer) [inline dask/array/core.py]: I think this does the wrong thing now. Previously, `asarray` would not convert an HDF5 dataset (for example) into a NumPy array but now it does.

Comment (mrocklin) [inline dask/array/core.py]: To be clear you're talking about the change of the predicate from "hasattr shape" to "shape is an iterable"?  If so then can you describe more about what an h5py.Dataset shape looks like?

Do we not test that h5py.Datasets stay h5py.Datasets until necessary?

Comment (shoyer) [inline dask/array/core.py]: Yes. The point was that `h5py.Dataset` defined a `.shape` attribute, which indicates that it was already an "array-like" object. So it seemed like the right thing is probably to not eagerly evaluate it -- that was the entire point of this check.

I'm not sure we actually have a test for this functionality, though, so it's totally expected that it broke.

Comment (mrocklin) [inline dask/array/core.py]: Do you have a recommended fix?

Comment (shoyer) [inline dask/array/core.py]: I would just stick with my original logic here, but it looks like that must have caused issues for #1151?

**Comment by mrocklin** [comment]
Indeed.  I'm still not sure I understand the h5py issue.

Datasets lazily create a shape attribute, so hasattr returns False while
getattr returns something?  If this understanding is correct then we could
probably come up with a more direct check.  I may be mistaken on what's
going on there.

I think the goal is for this check to fail on Dataset objects, but pass on
things like numpy arrays.  Is there a check that is more in line with
objectives?

**Comment by shoyer** [comment]
@mrocklin I was looking for a check that *passes* on `h5py.Dataset` (should not be coerced to numpy array) but fails on `list` (should be coerced to numpy array). `h5py.Dataset.shape` *is* defined.

**Comment by mrocklin** [comment]
Is it not an iterable?  Why does it fail the current check?

**Comment by shoyer** [comment]
Oops: got passes/fails backwards in my last comment. Edited.

**Comment by mrocklin** [comment]
Things that would change under this test are objects that have a shape attribute, but where that shape attribute is not an iterable.  For anything array-like (shape is a tuple) I definitely would not expect a change in behavior.

**Comment by shoyer** [comment]
For h5py.Dataset:

- `not hasattr(array, 'shape')` -> `False`
- `if isinstance(getattr(array, 'shape', None), Iterable)` -> `True`

**Comment by mrocklin** [comment]
Ah, so perhaps we're actually just missing a `not` here?

**Comment by mrocklin** [comment]
Anyway, if you can find the time to craft a suitable test I'll endeavor to find th etime to screw around with source.

**Pull request merged**